src/service.js

```diff
@@ -607,20 +684,7 @@ module.exports = function(mixinOptions) {
 					this.apolloServer = new ApolloServer({
 						schema,
 						..._.defaultsDeep({}, mixinOptions.serverOptions, {
-							context: ({ req, connection }) => ({
-								...(req
-									? {
-											ctx: req.$ctx,
-											service: req.$service,
-											params: req.$params,
-									  }
-									: {
-											ctx: connection.context.$ctx,
-											service: connection.context.$service,
-											params: connection.context.$params,
-									  }),
-								dataLoaders: new Map(), // create an empty map to load DataLoader instances into
-							}),
+							context: integrationContext => this.actions.context(integrationContext),
```

Comment:
Forgive my lack of knowledge with subscription behaviors, so I'll probably ask questions along the way here.

Does this specific change imply that a fresh moleculer context will be created each time a subscription does to resolve, or is this the moleculer context for the entire lifespan of the connected subscription?

As far as I can tell, the primary reason for the `ws` action was to generate a moleculer context at connection start and bind various data to it.  Presumably, that context was for the entire lifespan of the connected subscription.  I'm trying to understand if the generation of moleculer context is redundant.  That is, could one or the other of these calls to `this.actions.xxx` be handled instead with a plain old method call, or is there a benefit to establishing new moleculer Context for each of these independently?

Reply:
No problem, it's taking me a bit to really wrap my head around all this too, and I may be wrong, so I'm more than glad to do some mental gymnastics to make sure we get this right. 

If you haven't caught up or had the opportunity recently to read my original [issue](https://github.com/moleculerjs/moleculer-apollo-server/issues/75) You may want to do that because I go over my reasoning for a lot of this and the problems I'm trying to solve.

Apollo will create a new context for for every GQL **operation**.  So you can actually connect, and do nothing. So the `onConnect` hook is important for people who want to guard that entry point. Once an operation, of any kind, is sent via WebSocket, or HTTP, then it will get a new context for that operation. We use WebSockets as our main messaging transport, we open the connection and reuse it not only for subscriptions, but also for all queries and mutations.

Reply:
OK, so if I am understanding this correctly, the lifecycle is as follows:

* `onConnect` runs when the client connects.  This runs a single time for the entire subscription lifecycle.
* A new _Apollo_ `Context` is created each time a subscription publishes a new event, or a query/mutation is performed.  This is represented in the `context: integrationContext => this.actions.context(integrationContext)` line.

So in each of these scenarios, an action call is being used which is establishing a new _moleculer_ `Context`.  In addition, this will allow consumers to wrap those actions with various moleculer lifecycle handlers if they so desire.

What I think I am still missing then is the need for the `createAsyncIteratorContext` method getting called.  If a fresh _Apollo_ context and a fresh _moleculer_ context are both getting created during each operation, I'm guessing I'm misinterpreting what is considered an _operation_ as it pertains to subscriptions and that does not include the publish event.

Assuming that I am misinterpreting this and a fresh _moleculer_ context is needed each time a publish event is fired, I'm concerned about a few things:
1. The DataLoader cache.  So I'll fill in a little more here about what I'm concerned about here.  DataLoader instances are lazily loaded into the _Apollo_ Context as they become needed.  That is, every time a unique combination of action, args, and staticParams is detected, a new DataLoader instance is added to the `Map` inside the _Apollo_ Context.  This allows operations to be deferred and batched to limit the total number of calls during child resolution, provided that the action being called complies with the rules of DataLoader (must accept an array, must return results sequenced with that array, the output array must be the same length of the input array).  One thing DataLoader does is cache results for a given "id".  Therefore, if you request the same id it will return the Promise that was bound to the previous time that "id" was requested.  This limits the total number of ids that need to be called.  However, this cache is intended to be short lived and essentially last for a single request's lifespan.  If it lasted longer than that, the cache would grow stale and potentially be out of date.  There are cache invalidation mechanisms in DataLoader, but they generally just advise to create new instances on each request rather than trying to engage with the cache invalidation.  If the _Apollo_ Context is long-lived for a subscription, then that would also imply that DataLoader instances are long lived, which would result in stale caches.  If the _Apollo_ Context is short-lived then that isn't a concern, but its likely a mechanism to clear the DataLoader `Map` in some way (potentially at publish event time?) would be needed if the Apollo Context isn't refreshed.
2. The use of `this.broker.ContextFactory.create` to generate a new _moleculer_ Context is, AFAIK, going to miss out on some things like the merging of `meta`.  If the `context` or `ws` actions are enriching the _moleculer_ Context with `meta` that is necessary for upstream actions, then that would cause an issue with those upstream operations expecting that `meta` to be present.  Again, assuming that we need a new moleculer Context for each publish event, it might be preferable to make use of an action handler (similar to the `context` and `ws` actions) to generate the context so that it facilitates this merging properly.  Alternatively, it appears to me that the `Context.create` method accepts an option for `parentCtx`, which it would utilize to do the merging of things like `meta`.

Reply:
@NickClark I sent you a message on Discord to see if we can schedule some time to discuss this.  It would probably be more efficient if we can do some interactive back and forth rather than trying to hash things out with this delay.  Let me know if that works for you.